### PR TITLE
1853 id verification consent sms

### PR DESCRIPTION
**Pull request by eastandwestwind**
Closes https://github.com/ethyca/fides/issues/1853

### Code Changes

* [ ] Adjust back-end to support both email and SMS id verification for consent
* [ ] Adds SMS input field to privacy center
* [ ] Adds unit tests

### Steps to Confirm

* [ ] In `fides.toml` set `subject_identity_verification_required = true` and `notification_service_type = "twilio_text"`
* [ ] `nox -s dev`
* [ ] Add twilio sms config/secret using postman (follow https://docs.ethyca.com/fides/dsr_quickstart/dsr_support/messaging#twilio to set up your twilio acct)
* [ ] Ensure your phone number has been added to the twilio account's verified phone numbers (https://console.twilio.com/us1/develop/phone-numbers/manage/verified)
* [ ] In a new terminal, `cd` into `clients/privacy-center`, run `npm install`, `npm run build`, then `npm run dev`
* [ ] Submit consent request, enter verification code sent to phone
* [ ] Please also regression test email id verification for consent

### Pre-Merge Checklist

* [ ] All CI Pipelines Succeeded
* Documentation:
  * [x] documentation complete, [PR opened in fidesdocs](https://github.com/ethyca/fidesdocs/pulls)
  * [ ] documentation [issue created in fidesdocs](https://github.com/ethyca/fidesdocs/issues/new/choose)
* [x] Issue Requirements are Met
* [ ] Relevant Follow-Up Issues Created
* [x] Update `CHANGELOG.md`

### Description Of Changes

_Write some things here about the changes and any potential caveats_

### CODE REVIEW

File: src/fides/api/ops/api/v1/endpoints/consent_request_endpoints.py [hunk 1]
```diff
@@ -69,26 +72,46 @@ def create_consent_request(
...
+                identity_data=data,
+                identity_type=ProvidedIdentityType.phone_number.value,
+            )
+        else:
+            identity = _get_or_create_provided_identity(
+                db=db,
+                identity_data=data,
+                identity_type=ProvidedIdentityType.email.value,
+            )
+    elif data.email:
```
Comment (adamsachs) [inline src/fides/api/ops/api/v1/endpoints/consent_request_endpoints.py:105]: this may be a bit deeper than we want to go in this PR, but i think there's an edge case here that i'm not sure we're handling very cleanly. specifically, if the client (whether it's the privacy center UI or theoretically some other API client) here submits _only_ an `email` or `phone_number` while their app is configured for an incompatible notification service  -- e.g. if their `notification_service_type` property is set to `twilio_text` but the request here contains only an `email`.

i think/assume we'll probably end up erroring out much further downstream in the messaging dispatcher - but i think we could probably be a bit more defensive here (and any other endpoints that go through similar dispatching logic), and just cut things off before we get that far to give what's likely a clearer error message to the user (probably a `422` here?)

but maybe i'm missing something. either way, i don't think it's a must-have at the moment, but curious to get your thoughts.

File: src/fides/api/ops/api/v1/endpoints/consent_request_endpoints.py [hunk 2]
```diff
@@ -69,26 +72,46 @@ def create_consent_request(
...
-        }
-        identity = ProvidedIdentity.create(db, data=provided_identity_data)
+    # consent requests, unlike privacy requests, only accept 1 identity type- email or phone
+    # if both identity types are provided, use identity type if defined in CONFIG.notifications.notification_service_type, else default to email
+    if data.email and data.phone_number:
+        messaging_method = get_messaging_method(
+            CONFIG.notifications.notification_service_type
+        )
+        if messaging_method == MessagingMethod.EMAIL:
+            identity = _get_or_create_provided_identity(
```
Comment (adamsachs) [inline src/fides/api/ops/api/v1/endpoints/consent_request_endpoints.py:88]: a bit of a nit, but i wonder if it'd make things a bit cleaner to move the `get_messaging_method` check and the different branches here _within_ the `_get_or_create_provided_identity` utility function. it may lead to a bit less code, and it also may make the top-level endpoint method here a bit more readable - all the details about getting the `identity` instance would be fully abstracted away from the top-level method, which you've already started to do here (and i think is a good direction to go in generally).

that being said, it's not too hard to follow as is, so not a huge deal. it just feels like some of the conditional branching here could be moved out of the top-level function, and possibly consolidated a bit if done more closely to the identity retrieval.

**Review by adamsachs — APPROVED** [review]
the BE here generally looks good to me, pretty clean updates! just a couple of minor points that i don't think are _necessary_ to address, but wanted to call attention to them.

a couple of other asides that i noticed along the way, which we may want to create follow ups for:
- i found the setup [docs](https://docs.ethyca.com/fides/dsr_quickstart/dsr_support/messaging#twilio-sms-config) generally good, but lacking a crucial detail that indicates the actual URL/path that needs to be hit for the given steps. am i missing something obvious? there do seem to be other points in the docs where we've also omitted the URL/path, so perhaps this is a deliberate decision? from the POV of a client, i think i'd find this pretty confusing. what do you think?
- i also noticed what seems to be a bug in our caching of verification codes - i don't think we're actually setting our verification codes to expire in the cache using the `identity_verification_code_ttl_seconds` setting, i think we're just relying on the `default_ttl_seconds` expiration time. i'll create a follow up ticket for this one and can probably get up a PR pretty quick to fix, just leaving a note here.

File: src/fides/api/ops/api/v1/endpoints/consent_request_endpoints.py (hunk 1 above)
Comment (eastandwestwind) [reply src/fides/api/ops/api/v1/endpoints/consent_request_endpoints.py:105]: @adamsachs you're correct that in this use case:

1. A customer wants to enable notifications
2. Their `notification_service_type` set to `twilio_text`
3. Their privacy-center `config.json` allows for only email to be sent via the API
4. A privacy request OR consent request is submitted with only an email identity
5. The app will raise a `500` with message ` Error sending the verification code message: No phone identity supplied.`. 

I'd prefer to keep this logic within the messaging service, since it's a central place to handle all things messaging and erroring. Otherwise, we'll have to implement the checks on each caller, which will be more of a hassle to manage (e.g. if we implement a new caller, we have to remember to implement this check).

For now, I'll open a new follow-up ticket to do the following:

1. move this check earlier on in the messaging service, and
2. add new Exception type to the dispatcher to raise `422` code instead, using this new type to handle this err case, and 
3. catch this new exception in every place where we call our message dispatcher, and
2. update the message to be clearer: ` Error sending the verification code message: Message service expecting sms identity, but no sms identity was supplied.

File: src/fides/api/ops/api/v1/endpoints/consent_request_endpoints.py (hunk 2 above)
Comment (eastandwestwind) [reply src/fides/api/ops/api/v1/endpoints/consent_request_endpoints.py:88]: hey, this is a good point, I'll give it a go!

File: src/fides/api/ops/api/v1/endpoints/consent_request_endpoints.py (hunk 1 above)
Comment (eastandwestwind) [reply src/fides/api/ops/api/v1/endpoints/consent_request_endpoints.py:105]: ticket here- https://github.com/ethyca/fides/issues/2106

**Comment by eastandwestwind** [comment]
Thanks so much for the thorough testing and review @adamsachs ! A couple things:

1. Based on your feedback, I've updated comments in the code to reflect he configurability of the PC: 
```
Consent requests, unlike privacy requests, only accept 1 identity type- email or phone. 
    These identity types are configurable as optional/required within the privacy center config.json.
    If both identity types are provided, we'll use identity type if defined in
    CONFIG.notifications.notification_service_type, else default to email.
```
2. As discussed in Slack, now that I know where the new docs live, I'll make the appropriate updates to Twilio configuration docs for clarity- https://github.com/ethyca/fides/issues/2107
3. Follow-up ticket for exception handling- https://github.com/ethyca/fides/issues/2106
4. Thanks so much for working on https://github.com/ethyca/fides/pull/2105, I'll take a look today!

File: src/fides/api/ops/api/v1/endpoints/consent_request_endpoints.py (hunk 1 above)
Comment (adamsachs) [reply src/fides/api/ops/api/v1/endpoints/consent_request_endpoints.py]: nice, your point about keeping the logic within the messaging service does make sense. and so moving the check earlier in the messaging service makes a lot of sense to me.

it's a detail, but i'm not sure if i like raising the `422` from within the dispatch service - but i do get that you want to distinguish it as a specific type of exception and not just a generic `MessageDispatchException`, so you can then handle it appropriately in the places that use the dispatcher (which i also think is a great approach). so maybe just another subtype of `MessageDispatchException` that clearly indicates what's going wrong here, so it can be properly handled by the endpoint code and reported as a `422`? i guess i'm just a bit worried that if we introduce `HTTPException`s into the `message_dispatch_service`, it breaks the separation of concerns. feels a bit cleaner to have the underlying service just describe what went wrong, independent of the HTTP endpoint and how it's called; and then the specific pieces of endpoint code figure out which `HTTPException` to throw back to the user based on what's thrown up by the service underneath.

anyway, i'm splitting hairs a bit here, so feel free to ignore :) in general i think where you landed makes a lot of sense!

Comment (eastandwestwind) [reply src/fides/api/ops/api/v1/endpoints/consent_request_endpoints.py]: ah I didn't mean literally raise the 422 as an `HTTPException`, but instead use a different error class, maybe `MessageDispatchValueException` so that the endpoint can handle throwing 422. 

But I like the idea of using a subtype of `MessageDispatchException` instead of a new Exception class altogether.

Comment (eastandwestwind) [reply src/fides/api/ops/api/v1/endpoints/consent_request_endpoints.py]: But yes, you're totally right! We shouldn't be involving HTTP concepts in the messaging service

Comment (eastandwestwind) [reply src/fides/api/ops/api/v1/endpoints/consent_request_endpoints.py]: updated requirements in ticket `add new Exception type (perhaps a subtype of `MessageDispatchException`) to the dispatcher such that the endpoint can catch and raise 422 code`

Comment (adamsachs) [reply src/fides/api/ops/api/v1/endpoints/consent_request_endpoints.py]: ah ok great, we're totally on the same page then  👍

**Comment by adamsachs** [comment]
nice, this all looks good to me on the backend pieces  - thanks for addressing everything i brought up in a thoughtful way!

not sure if we want to go ahead and merge without further FE code review? things certainly seem functional, but can't comment much on the code itself :)
